presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java

```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
         return ImmutableSet.copyOf(metastore.listRoles());
     }
 
+    @Override
+    public Set<RoleGrant> listAllRoleGrants(ConnectorSession session, Optional<Set<String>> roles, Optional<Set<String>> grantees, OptionalLong limit)
+    {
+        /*
+         * The Hive Metadata interface provides us with two ways to retrieve role grants:
+         * 1. principals by single role
+         * 2. roles by single principal
+         *
+         * We check the passed predicates to decide which path is most optimal.
+         */
+        if (grantees.isPresent()) {
+            if (roles.isPresent() && roles.get().size() <= grantees.get().size() * 2) {
+                // 2x because for the grantees we do not know the grantee type (ROLE or USER)
+                return getRoleGrantsByRoles(roles.get(), limit);
+            }
+            else {
+                return getRoleGrantsByGrantees(grantees.get(), limit);
+            }
+        }
+        else {
+            // if roles are present use those, otherwise get a list of roles from the metastore
+            return getRoleGrantsByRoles(roles.orElseGet(() -> metastore.listRoles()), limit);
+        }
+    }
+
+    private Set<RoleGrant> getRoleGrantsByGrantees(Set<String> grantees, OptionalLong limit)
+    {
+        ImmutableSet.Builder<RoleGrant> builder = ImmutableSet.builder();
+        int count = 0;
+        for (String grantee : grantees) {
+            for (PrincipalType type : new PrincipalType[]{USER, ROLE}) {
+                for (RoleGrant grant : metastore.listRoleGrants(new HivePrincipal(type, grantee))) {
+                    if (limit.isPresent() && count >= limit.getAsLong()) {
```

Comment:
you should verify limit before doing a call to HMS (in the middle loop)

Reply:
Same as above... A bit embarrassing.